core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -0,0 +1,214 @@
+package org.infinispan.container.versioning.irac;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import org.infinispan.commands.CommandsFactory;
+import org.infinispan.commands.irac.IracCleanupTombstoneCommand;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.distribution.DistributionInfo;
+import org.infinispan.distribution.DistributionManager;
+import org.infinispan.factories.KnownComponentNames;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.metadata.impl.IracMetadata;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.impl.VoidResponseCollector;
+import org.infinispan.util.ExponentialBackOff;
+import org.infinispan.util.concurrent.AggregateCompletionStage;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.infinispan.xsite.XSiteBackup;
+import org.infinispan.xsite.irac.DefaultIracManager;
+import org.infinispan.xsite.irac.IracExecutor;
+
+@Scope(Scopes.NAMED_CACHE)
+public class DefaultIracTombstoneManager implements IracTombstoneManager {
+
+   @Inject DistributionManager distributionManager;
+   @Inject RpcManager rpcManager;
+   @Inject CommandsFactory commandsFactory;
+   private final Map<Object, TombstoneData> tombstoneMap;
+   private final IracExecutor iracExecutor;
+   private final Collection<XSiteBackup> asyncBackups;
+
+   public DefaultIracTombstoneManager(Configuration configuration) {
+      this.iracExecutor = new IracExecutor(this::performCleanup);
+      this.asyncBackups = DefaultIracManager.asyncBackups(configuration);
+      this.tombstoneMap = new ConcurrentHashMap<>();
+   }
+
+   @Inject
+   public void inject(@ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService executorService,
+                      @ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor blockingExecutor) {
+      // using the inject method here in order to decrease the class size
+      iracExecutor.setBackOff(ExponentialBackOff.NO_OP);
+      iracExecutor.setExecutor(blockingExecutor);
+      // TODO configure?
```

Comment:
created: https://issues.redhat.com/browse/ISPN-13446

I'm not sure about the caffeine internals but it will block the "evict listener". Before removing the tombstone from the map, it requires to check all sites. Not sure if this blocking operation is fine.
I believe it would be better to offload to disk to reduce the memory pressure.

Reply:
We have a way to activate/passivate entries without blocking Caffeine, the same approach might work here. But eviction to disk is also an option.